lint/param-duplication_checker.go

```diff
@@ -2,79 +2,67 @@ package lint
 
 import (
 	"go/ast"
-	"strings"
 
 	"github.com/Quasilyte/astcmp"
 )
 
 func paramDuplicationCheck(ctx *context) func(*ast.File) {
-	return wrapParamListChecker(&paramDuplicationChecker{
-		baseParamListChecker: baseParamListChecker{ctx: ctx},
+	return wrapFuncDeclChecker(&paramDuplicationChecker{
+		baseFuncDeclChecker: baseFuncDeclChecker{ctx: ctx},
 	})
 }
 
 type paramDuplicationChecker struct {
-	baseParamListChecker
-
-	cause *ast.FuncDecl
+	baseFuncDeclChecker
 }
 
-func (c *paramDuplicationChecker) PerFuncInit(fn *ast.FuncDecl) bool {
-	c.cause = fn
-	return true
+func (c *paramDuplicationChecker) CheckFuncDecl(decl *ast.FuncDecl) {
+	opt := c.optimiseFuncType(decl.Type)
+	if !astcmp.EqualExpr(opt, decl.Type) {
+		c.warn(decl.Type, opt)
+	}
 }
 
-func (c *paramDuplicationChecker) CheckParamList(params []*ast.Field) {
-	if len(params) < 2 {
-		return
-	}
-	equalRange := [][]*ast.Field{}
-	equalRange = append(equalRange, []*ast.Field{params[0]})
-	warn := false
-	for i, p := range params[1:] {
-		if astcmp.EqualExpr(p.Type, params[i].Type) {
-			equalRange[len(equalRange)-1] =
-				append(equalRange[len(equalRange)-1], p)
-			warn = true
-		} else {
-			equalRange = append(equalRange, []*ast.Field{p})
-		}
-	}
-	if warn {
-		c.warn(equalRange)
+func (c *paramDuplicationChecker) optimiseFuncType(f *ast.FuncType) *ast.FuncType {
+	res := c.optimiseParams(f.Results)
+	par := c.optimiseParams(f.Params)
+	return &ast.FuncType{
+		Params:  par,
+		Results: res,
 	}
 }
+func (c *paramDuplicationChecker) optimiseParams(params *ast.FieldList) *ast.FieldList {
+	if params == nil || len(params.List) < 2 {
+		return params
+	}
 
-func (c *paramDuplicationChecker) warn(fields [][]*ast.Field) {
-	paramsBefore := []string{}
-	paramsAfter := []string{}
-	for _, flist := range fields {
-		typeBefore := []string{}
-		typeAfter := []string{}
-
-		for _, f := range flist {
-			typeBefore = append(typeBefore,
-				c.paramNamesStr(f.Names)+" "+nodeString(c.ctx.FileSet, f.Type))
-			typeAfter = append(typeAfter,
-				c.paramNamesStr(f.Names))
+	newParams := &ast.FieldList{
+		List: []*ast.Field{},
+	}
+	names := make([]*ast.Ident, len(params.List[0].Names))
+	copy(names, params.List[0].Names)
+	newParams.List = append(newParams.List, &ast.Field{
```

Comment:
(optional)
It may help readability to assign `newParams.List` to `list` variable.
Then you can use it everywhere instead of `newParams.List`.
And in the end of the function you assign `list` to `newParams.list`.'